File: keenetic.py
import requests
import hashlib
import json
import os
from dotenv import load_dotenv

# Загрузка переменных окружения
load_dotenv()

# --- Настройки ---
ROUTER_IP = os.getenv("ROUTER_IP")
LOGIN = os.getenv("LOGIN")
PASSWORD = os.getenv("PASSWORD")
COOKIES_FILENAME = os.getenv("COOKIES_FILE", "auth_cookies.json")
DATA_DIR = "data"
COOKIES_FILE = os.path.join(DATA_DIR, COOKIES_FILENAME)

def get_md5(data):
    return hashlib.md5(data.encode('utf-8')).hexdigest()

def get_sha256(data):
    return hashlib.sha256(data.encode('utf-8')).hexdigest()

def save_cookies(session, filename):
    try:
        # Убедимся, что папка существует
        os.makedirs(os.path.dirname(filename), exist_ok=True)
        with open(filename, 'w') as f:
            json.dump(session.cookies.get_dict(), f)
        print(f"Куки сохранены в {filename}")
    except Exception as e:
        print(f"Ошибка сохранения куки: {e}")

def load_cookies(session, filename):
    if os.path.exists(filename):
        try:
            with open(filename, 'r') as f:
                cookies = json.load(f)
                session.cookies.update(cookies)
            print(f"Куки загружены из {filename}")
            return True
        except Exception as e:
            print(f"Ошибка загрузки куки: {e}")
    return False

def authenticate():
    session = requests.Session()
    
    # Попытка восстановить сессию
    if load_cookies(session, COOKIES_FILE):
        print("Проверка сохраненной сессии...")
        try:
            # Проверяем валидность сессии запросом информации о системе
            check_url = f"http://{ROUTER_IP}/rci/show/system"
            r = session.get(check_url, timeout=5)
            if r.status_code == 200:
                print("Сессия валидна.")
                return session
            else:
                print(f"Сессия устарела (код {r.status_code}). Требуется повторная авторизация.")
        except Exception as e:
            print(f"Ошибка при проверке сессии: {e}")

    auth_url = f"http://{ROUTER_IP}/auth"

    # 1. Первый GET запрос для получения соли (Challenge)
    print("--- Шаг 1: Получение Challenge ---")
    r1 = session.get(auth_url)
    
    challenge = r1.headers.get('X-NDM-Challenge')
    realm = r1.headers.get('X-NDM-Realm')
    product = r1.headers.get('X-NDM-Product')
    
    print(f"X-NDM-Challenge: {challenge}")
    print(f"X-NDM-Realm: {realm}")
    print(f"X-NDM-Product: {product}")

    if not challenge or not realm:
        print("Ошибка: не удалось получить данные для авторизации.")
        return None

    # 2. Вычисление хеша
    # Формула: SHA256(challenge + MD5(login + ":" + realm + ":" + password))
    h1_string = f"{LOGIN}:{realm}:{PASSWORD}"
    h1 = get_md5(h1_string)
    print(f"MD5({h1_string}): {h1}")
    
    final_hash = get_sha256(challenge + h1)
    print(f"\nВычисленный хеш: {final_hash}")

    # 3. POST запрос для авторизации
    print("\n--- Шаг 2: Отправка хеша ---")
    auth_data = {
        "login": LOGIN,
        "password": final_hash
    }
    
    r2 = session.post(auth_url, json=auth_data)
    
    print(f"Статус ответа: {r2.status_code}")
    
    # Проверяем куки (Session ID сохраняется в них автоматически в объекте session)
    if r2.status_code == 200:
        print("Авторизация успешна!")
        print("Полученные куки (сессия):", session.cookies.get_dict())
        save_cookies(session, COOKIES_FILE)
        return session
    else:
        print("Ошибка авторизации!")
        return None

# --- Пример использования ---
if __name__ == "__main__":
    authorized_session = authenticate()
    
    if authorized_session:
        # Пробуем получить информацию о системе
        print("\n--- Шаг 3: Тестовый запрос (show system) ---")
        api_url = f"http://{ROUTER_IP}/rci/show/system"
        response = authorized_session.get(api_url)
        
        if response.status_code == 200:
            data = response.json()
            print(f"Модель устройства: {data.get('model')}")
            print(f"Версия ОС: {data.get('release')}")
            print(f"Аптайм: {data.get('uptime')} сек.")
        else:
            print(f"Ошибка RCI: {response.status_code}")

        # Получение списка клиентов (show ip hotspot)
        print("\n--- Шаг 4: Список клиентов (show ip hotspot) ---")
        hotspot_url = f"http://{ROUTER_IP}/rci/show/ip/hotspot"
        resp_hotspot = authorized_session.get(hotspot_url)

        if resp_hotspot.status_code == 200:
            clients = resp_hotspot.json()
            # Ответ может быть списком или объектом, содержащим список. Обычно это список объектов.
            # На всякий случай проверим структуру, но обычно это просто список хостов.
            
            # В некоторых версиях API ответ может быть обернут.
            # Если clients это словарь (dict), попробуем найти в нем список.
            if isinstance(clients, dict):
                 # Иногда бывает так: {"host": [...]}
                 clients = clients.get('host', [])
            
            # Фильтруем только активных клиентов
            active_clients = [c for c in clients if c.get('active')]

            print(f"Найдено активных клиентов: {len(active_clients)}")
            print("-" * 60)
            print(f"{'IP Address':<16} | {'MAC Address':<18} | {'Hostname'}")
            print("-" * 60)
            
            for client in active_clients:
                ip = client.get('ip', 'N/A')
                mac = client.get('mac', 'N/A')
                name = client.get('name', '')
                hostname = client.get('hostname', '')
                if not name:
                    name = client.get('hostname', 'Unknown')
                
                print(f"{ip:<16} | {mac:<18} | {hostname}")
        else:
            print(f"Ошибка получения списка клиентов: {resp_hotspot.status_code}")

     

File: keenetic.js
require('dotenv').config();
const axios = require('axios');
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');
const { logger } = require('./logger');

// --- Настройки ---
const ROUTER_IP = process.env.ROUTER_IP;
const LOGIN = process.env.LOGIN;
const PASSWORD = process.env.PASSWORD;
const COOKIES_FILENAME = process.env.COOKIES_FILE || 'auth_cookies.json';
const DATA_DIR = 'data';
const COOKIES_FILE = path.join(__dirname, DATA_DIR, COOKIES_FILENAME);

// --- Утилиты ---
const getMd5 = (data) => crypto.createHash('md5').update(data).digest('hex');
const getSha256 = (data) => crypto.createHash('sha256').update(data).digest('hex');

class KeeneticClient {
    constructor() {
        this.cookies = {};
        this.client = axios.create({
            baseURL: `http://${ROUTER_IP}`,
            timeout: 5000,
            validateStatus: () => true, // Не выбрасывать ошибку на 4xx/5xx
        });

        // Интерцептор для добавления куки к запросам
        this.client.interceptors.request.use((config) => {
            const cookieString = Object.entries(this.cookies)
                .map(([key, value]) => `${key}=${value}`)
                .join('; ');
            if (cookieString) {
                config.headers['Cookie'] = cookieString;
            }
            return config;
        });

        // Интерцептор для сохранения куки из ответов
        this.client.interceptors.response.use((response) => {
            if (response.headers['set-cookie']) {
                response.headers['set-cookie'].forEach(cookieStr => {
                    const [cookie] = cookieStr.split(';');
                    const [key, value] = cookie.split('=');
                    if (key && value) {
                        this.cookies[key.trim()] = value.trim();
                    }
                });
                // Если куки обновились, можно сохранить их (опционально здесь, или явно после авторизации)
            }
            return response;
        });
    }

    loadCookies() {
        if (fs.existsSync(COOKIES_FILE)) {
            try {
                const data = fs.readFileSync(COOKIES_FILE, 'utf8');
                this.cookies = JSON.parse(data);
                logger.debug(`Куки загружены из ${COOKIES_FILE}`);
                return true;
            } catch (e) {
                logger.debug(`Ошибка загрузки куки: ${e.message}`);
            }
        }
        return false;
    }

    saveCookies() {
        try {
            if (!fs.existsSync(path.dirname(COOKIES_FILE))) {
                fs.mkdirSync(path.dirname(COOKIES_FILE), { recursive: true });
            }
            fs.writeFileSync(COOKIES_FILE, JSON.stringify(this.cookies, null, 2));
            logger.info(`Куки сохранены в ${COOKIES_FILE}`);
        } catch (e) {
            logger.error(`Ошибка сохранения куки: ${e.message}`);
        }
    }

    async authenticate() {
        // 1. Попытка восстановить сессию
        if (this.loadCookies()) {
            logger.debug("Проверка сохраненной сессии...");
            try {
                const res = await this.client.get('/rci/show/system');
                if (res.status === 200) {
                    logger.debug("Сессия валидна.");
                    return true;
                } else {
                    logger.debug(`Сессия устарела (код ${res.status}). Требуется повторная авторизация.`);
                }
            } catch (e) {
                logger.debug(`Ошибка при проверке сессии: ${e.message}`);
            }
        }

        // 2. Начало новой авторизации
        logger.debug("--- Шаг 1: Получение Challenge ---");
        const r1 = await this.client.get('/auth');

        const challenge = r1.headers['x-ndm-challenge'];
        const realm = r1.headers['x-ndm-realm'];
        const product = r1.headers['x-ndm-product'];

        logger.debug(`X-NDM-Challenge: ${challenge}`); // Changed to debug
        logger.debug(`X-NDM-Realm: ${realm}`); // Changed to debug
        logger.debug(`X-NDM-Product: ${product}`); // Changed to debug

        if (!challenge || !realm) {
            logger.error("Ошибка: не удалось получить данные для авторизации.");
            return false;
        }

        // 3. Вычисление хеша
        // Формула: SHA256(challenge + MD5(login + ":" + realm + ":" + password))
        const h1String = `${LOGIN}:${realm}:${PASSWORD}`;
        const h1 = getMd5(h1String);
        logger.debug(`MD5(${h1String}): ${h1}`); // Changed to debug

        const finalHash = getSha256(challenge + h1);
        logger.debug(`\nВычисленный хеш: ${finalHash}`); // Changed to debug

        // 4. Отправка хеша
        logger.debug("\n--- Шаг 2: Отправка хеша ---");
        const authData = {
            login: LOGIN,
            password: finalHash
        };

        const r2 = await this.client.post('/auth', authData);
        logger.debug(`Статус ответа: ${r2.status}`);

        if (r2.status === 200) {
            logger.debug("Авторизация успешна!");
            logger.debug(`Полученные куки (сессия): ${JSON.stringify(this.cookies)}`);
            this.saveCookies();
            return true;
        } else {
            logger.error("Ошибка авторизации!");
            return false;
        }
    }

    async getSystemInfo() {
        logger.debug("--- Шаг 3: Тестовый запрос (show system) ---");
        const res = await this.client.get('/rci/show/system');
        if (res.status === 200) {
            const data = res.data;
            // logger.debug(`DEBUG: ${JSON.stringify(data, null, 2)}`); // Uncomment for debug
            logger.debug(`SYSTEM INFO: Hostname: ${data.hostname}\t Uptime: ${data.uptime} сек.`);
            // logger.info(`Модель устройства: ${data.hostname}`);
            // logger.info(`Версия ОС: ${data.release}`);
            // logger.info(`Аптайм: ${data.uptime} сек.`);
        } else {
            logger.error(`Ошибка RCI: ${res.status}`);
        }
    }

    async getHotspotClients() {
        logger.debug("--- Шаг 4: Список клиентов (show ip hotspot) ---");
        const res = await this.client.get('/rci/show/ip/hotspot');

        if (res.status === 200) {
            let clients = res.data;

            // Нормализация ответа
            if (clients && !Array.isArray(clients) && typeof clients === 'object') {
                clients = clients.host || [];
            }
            if (!Array.isArray(clients)) {
                clients = [];
            }

            const activeClients = clients.filter(c => c.active);

            logger.info(`Найдено активных клиентов: ${activeClients.length}`);
            logger.debug("-".repeat(60));
            logger.debug(`${"IP Address".padEnd(16)} | ${"MAC Address".padEnd(18)} | ${"Hostname"}`);
            logger.debug("-".repeat(60));

            activeClients.forEach(client => {
                const ip = client.ip || 'N/A';
                const mac = client.mac || 'N/A';
                let name = client.name || '';
                const hostname = client.hostname || '';
                if (!name) name = hostname || 'Unknown';

                logger.debug(`${ip.padEnd(16)} | ${mac.padEnd(18)} | ${name}`);
            });

            return activeClients;
        } else {
            logger.error(`Ошибка получения списка клиентов: ${res.status}`);
            return [];
        }
    }
}

module.exports = KeeneticClient;

// --- Запуск ---
// logger.info('KeeneticClient module loaded.');
// logger.info('require.main === module:', require.main === module);

if (require.main === module) {
    (async () => {
        if (!ROUTER_IP || !LOGIN || !PASSWORD) {
            logger.error("Ошибка: Не заданы переменные окружения ROUTER_IP, LOGIN, PASSWORD");
            process.exit(1);
        }

        const keenetic = new KeeneticClient();
        const isAuth = await keenetic.authenticate();

        if (isAuth) {
            await keenetic.getSystemInfo();
            await keenetic.getHotspotClients();
        }
    })();
}
